lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```diff
@@ -425,6 +438,202 @@ prop_availableUTxO makeProperty =
+        utxoEven :: UTxO
+        utxoEven = changeUTxO txSet Even
+
+        -- The UTxO set that contains only odd-parity output addresses.
+        utxoOdd :: UTxO
+        utxoOdd  = changeUTxO txSet Odd
+
+        txSet :: Set Tx
+        txSet = Set.fromList $ uncurry makeTx <$> (zip txIds txOutputs)
+
+        txIds :: [Hash "Tx"]
+        txIds = makeTxId <$> [minBound .. maxBound]
+
+    -- Creates a transaction from a transaction id and a set of outputs, by
+    -- adding dummy data for fields that are not used by 'changeUTxO'.
+    --
+    -- Ideally, we'd leave all of the unused fields undefined (to assert that
+    -- they should not be evaluated or processed in any way), but since the
+    -- fields of the 'Tx' type are strict, our next best option is to provide
+    -- a minimal value for each field.
+    --
+    makeTx :: Hash "Tx" -> TxOutputs -> Tx
+    makeTx txId TxOutputs {unTxOutputs} = Tx
+        { txId
+        , outputs = unTxOutputs
+        -- Unused fields:
+        , fee = Nothing
+        , resolvedCollateral = []
+        , resolvedInputs = []
+        , metadata = Nothing
+        , withdrawals = Map.empty
+        }
+
+    -- Creates a dummy transaction identifier from a single word.
+    makeTxId :: Word8 -> Hash "Tx"
+    makeTxId i = Hash $ BS.pack [i]
+
+-- | Represents all the outputs of a transaction (both change and non-change).
+--
+newtype TxOutputs = TxOutputs
+    { unTxOutputs :: [TxOut]
+        -- ^ A transaction's outputs.
+    }
+    deriving (Eq, Generic, Show)
+
+genTxOutputs :: Gen TxOutputs
+genTxOutputs = TxOutputs <$> listOf genTxOut
+
+shrinkTxOutputs :: TxOutputs -> [TxOutputs]
+shrinkTxOutputs (TxOutputs outs) = TxOutputs <$> shrinkList shrinkTxOut outs
+
+-- | An implementation of 'IsOurs' that matches all addresses.
+--
+instance IsOurs All Address where
+    isOurs = isOursIf (const True)
+
+data All = All
+
+-- | An implementation of 'IsOurs' that matches addresses by their parity.
+--
+instance IsOurs Parity Address where
+    isOurs a p = isOursIf ((== p) . addressParity) a p
+
+-- | Represents the parity of a value (whether the value is even or odd).
+--
+data Parity = Even | Odd
+    deriving (Eq, Show)
+
+-- | Computes the parity of an address.
+--
+-- Parity is defined in the following way:
+--
+--    - even-parity address:
+--      an address with a pop count (Hamming weight) that is even.
+--
+--    - odd-parity address:
+--      an address with a pop count (Hamming weight) that is odd.
+--
+-- Examples of even-parity and odd-parity addresses:
+--
+--    - 0b00000000 : even (Hamming weight = 0)
+--    - 0b00000001 : odd  (Hamming weight = 1)
+--    - 0b00000010 : odd  (Hamming weight = 1)
+--    - 0b00000011 : even (Hamming weight = 2)
+--    - 0b00000100 : odd  (Hamming weight = 1)
+--    - ...
+--    - 0b11111110 : odd  (Hamming weight = 7)
+--    - 0b11111111 : even (Hamming weight = 8)
+--
+addressParity :: Address -> Parity
+addressParity = parity . addressPopCount
```

Comment:
Any other `Address -> Bool` function would be just as good, as long it splits the set of generated addresses into roughly equal parts.

Reply:
Yes, definitely.

It's an arbitrary choice to use this particular function. It has two nice advantages:

1. Addresses are "even" and "odd" with equal frequency (meaning we don't have to do anything special with generation).
2. Address parity can be determined with a pure function (meaning we don't have to keep any state around in our `IsOurs` instance).

I've added a comment to `prop_changeUTxO` with these reasons.